Iot2000 ebg swu

andi8086 opened the pull request:
# EFIBootGuard and SWUpdate #

These commits update the existing `meta-iot2000` layer to use integrated `wic` and `kas` build tools.
The boot loader is changed to `EFI Boot Guard` and some fixes are applied. The 2nd half of the commits integrate `SWUpdate` to support system updates.

Commit:
meta-iot2000-bsp: Change bootloader to efibootguard

* In machine config set EFI_PROVIDER to efibootguard.
wic-image.inc then does not depend on systemd-boot anymore.
* Change default wks header to include config partition.
* Adapt boot parameters for efibootguard and select efibootguard
as loader.
* Change partition label to gpt.

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
Add meta-efibootguard to kas file

Change iot2000-example-image to include meta-efibootguard
in its kas file.

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
Add gptfdisk for wic

For wic to create a GPT, gptfdisk is required. Make the images depend
on it.

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
swupdate: Add swupdate using official meta layer

Import official meta-swupdate layer.
meta-swupdate needs some minor overrides to be compatible with our
update process and configuration.

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
Add native bg_setenv to the image dependencies

Efibootguard's tool is needed to create the config partition

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
wic: Export needed bitbake variables to wic

Add needed variables to WICVARS via a local.conf entry to have this
variable available in the wic source plugins.

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
Add update-image recipes and sw-description

Add sw-description as an update example and
an example recipes to build update images.

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
Be sure do_bootimg ends before do_image_wic starts

do_image_wic requires files that do_bootimg generates.

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
Improve expandfs script

* Add error handling.
* Add logging.
* Do not assume we only have three partitions. Always expand the last
partition of the root device.

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
Add some information about swupdate

Add information about how swupdate is integrated and explain the
example recipes.

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
swupdate: Disable backends and use rr LUA handler

Use round-robin LUA handler and don't use any backends.

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
Add efibootguard-tools to image

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
Add example partition layout for swupdate

* for image with two root file systems and system updates

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
Create and use /data for last partition

Last partition is a persistent storage partition that will be
mounted to /data.

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
swupdate: Add image recipes for full swupdate support

New image include file "example-image-swu.inc" includes partition
layout for swupdate (two root file systems and two config partitions
for efibootguard) and also adds swupdate to the image.

Two new image recipes with -swu suffix build existing images with
swupdate support.

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
Fix wic-image.inc

It does not make sense to first only apply ".wic" FS_TYPE
if the machine is 'iot2000' but then use the ${MACHINE} variable
for the wks file since the first append only allows one machine.

So remove the machine specific append and append for all machines.

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
Split common inc for wks file into generic parts

Split the common.iot2000.wks.inc into a generic
header and a generic footer part with a variable
part in the middle for partitions.

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

Commit:
Change rootfs for default wic image from ext3 to ext4

No technical reason to stay with ext3 so update.

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

stormc commented on meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh:
```diff
@@ -1,31 +1,29 @@
 #!/bin/sh
 
-set -eu
```
Why did you remove this? At least `-u` has some value.

andi8086 replied:
If something fails, for example sfdisk, the script should still disable itself that's why I removed set -e. -u does not help either... if some prog fails and we i.e. set ROOT_DEVICE=, then ROOT_DEVICE is no unbound variable anymore so that -u will not cause an exit of the shell. As the script is tested already, there are no misspellings either. So I decided to remove unneeded stuff.

stormc replied:
Hm, if e.g. sfdisk fails, then there should be some notification to the user or something. As is, it fails silently, doesn't it?

andi8086 replied:
I would say in our case, if sfdisk fails at that step, we have a problem with the whole system and it wont boot...

stormc replied:
Yes, granted. Nonetheless, there should be a note left to the user that and what failed.

stormc commented on meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh:
```diff
@@ -1,31 +1,29 @@
 
-parted -ms $ROOT_DEVICE rm $PART_NUMBER
+log "Resizing last partition"
+sfdisk -d ${ROOT_DEVICE} 2>/dev/null | grep -v last-lba | \
```
Are you sure the locale is actually EN or is, e.g., the DE variant also printing "last-lba"?

andi8086 replied:
"sudo LANG=de_DE.UTF-8 sfdisk -d ...." also prints "last-lba"

stormc commented on meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh:
```diff
@@ -1,31 +1,29 @@
 
-PART_NUMBER=3
+log () {
+	echo "${THIS_SCRIPT}: $1" > /dev/kmsg
```
Below, you use `logger`. Why not also here or just one of both?

andi8086 replied:
The step to be done is logged in the boot log and the detailed output of the programs is logged with logger. I thought this is a good idea to separate them for cleaner kernel logs.

stormc replied:
Hm, seems inconsistent to me to have to look in 2 places, particularly as this script is intended to run once. I'd prefer to have this in one central place.

andi8086 replied:
okay

stormc commented on meta-iot2000-example/recipes-example/iot2000setup/files/expandfs.sh:
```diff
@@ -1,31 +1,29 @@
+	sfdisk --force ${ROOT_DEVICE} 2>&1 | logger -t ${THIS_SCRIPT}
 
-parted -ms $ROOT_DEVICE unit s -- mkpart primary ext3 $START_BLOCK -1
+log "Informing kernel about new partitioning"
```
Doubly logged disabling, once here, once in the function.

andi8086 replied:
Oops

stormc commented on meta-iot2000-example/recipes-core/images/iot2000-example-image.inc:
```diff
@@ -43,3 +43,6 @@ IMAGE_INSTALL_append = " ofono ofono-tests"
 
 DEPENDS += "gptfdisk-native efibootguard-native"
+
+# The following is needed for correct EFI deployment
```
Yes, but the reason in code would be nice as well : )

andi8086 replied:
Sorry, what do you mean by that?

stormc replied:
You state it's needed. OK, but *why*, i.e., what's the reason it's needed? Jut to be verbose here...

andi8086 replied:
Ah sorry... I stumbled across the 'in code'. Since this is just the recipe :)

stormc commented on meta-iot2000-example/README.SWUPDATE.md:
```diff
@@ -0,0 +1,63 @@
+firmware and/or software updates targeting embedded devices.
+
+It can be run as a stand-alone server in foreground or as a daemon listening
+for connections via ethernet. Different backends can be configured to support
```
... ethernet or IPC

andi8086 replied:
true

stormc commented on meta-iot2000-example/README.SWUPDATE.md:
```diff
@@ -0,0 +1,63 @@
+
+
+The manual update process means, that the `swu` file must be copied onto the
+finished SD-card. To use it with swupdate, the following command is used:
```
.. or via SSH or something to the booted device. The setup can be adapted to work with backends like, e.g., hawkBit

andi8086 replied:
Added

stormc commented on meta-iot2000-example/recipes-support/swupdate/swupdate/swupdate.cfg:
```diff
@@ -0,0 +1,15 @@
+processes :
+(
+    {
+        name = "progress_firmware";
```
Well, there should be a word in the readme about this mechanism automatically rebooting after a firmware update.

andi8086 replied:
Added it in the last NOTE section.

stormc commented on meta-iot2000-example/recipes-support/swupdate/swupdate_%.bbappend:
```diff
@@ -1,5 +1,23 @@
+do_install_append() {
+        install -d ${D}${sysconfdir}
+        install -m 644 ${WORKDIR}/swupdate.cfg ${D}${sysconfdir}/swupdate.cfg
+        install -m 755 ${S}/tools/progress_firmware_unstripped ${D}/usr/bin/progress_firmware
```
Why the unstripped variant?

andi8086 replied:
This is because otherwise bitbake complains with QA issue that it was already stripped and will prevent future debugging. So it deploys the unstripped version in the recipe to strip it afterwards as I understand it.

stormc replied:
If bitbake strips it, you cannot sensibly debug anymore. So, it's a matter of correctly feeding bitbake's assumptions. I would prefer to selectively skip QA here or make a note why this is here for clarity. As is, one could also think of deploying unstripped binaries to a production target image...

andi8086 replied:
Okay I add an insane_skip

Commit:
util-linux: Update to v2.32.1

Use recipe and patches from current poky master
251ded90d4ad75da5e110e916068050f15ace07c

Remove virtual/crypt dependency which does not exist and seems to be not
needed.

Signed-off-by: Andreas Reichel <andreas.reichel.ext@siemens.com>

jan-kiszka reviewed (commented):
Do we really need to duplicate all files and patches that upstream already has for 2.30?

andi8086 commented:
> Do we really need to duplicate all files and patches that upstream already has for 2.30?

Yes we do. Of course patches don't apply anymore and the recipe itself is very different, as is the inc file. I have tried to use a bbappend before... but we need the newer patches, the newer inc file and some changes in the bb file. And also a bbappend wont have a correct version string in its name. It is a wonder that it builds at all... without warning or error.

jan-kiszka commented:
OK, different files need replacements, for sure. But there would generally be no need for bbappend hacks, see also d49d5dab3049.

Just checked: the majority of the files are identical, so please remove them. Moreover, the receipe should go into the example layer as that is the one which now needs the newer sfdisk.

andi8086 commented:
bbappends are no hacks but a general way to change things in an underlying recipe without duplicating files. If you say no duplication is wanted it is logical to speak of bbappends instead. Since bbappends are not possible and a lot has to be changed it is better to take the whole bunch because this is tested by upstream poky. I have the sha in the commit message where it comes from and that is poky maintained.

jan-kiszka commented:
Merged into master by now
